src/EFCore.PG/Query/EvaluatableExpressionFilters/Internal/NpgsqlHashingAlgorithmEvaluatableExpressionFilter.cs

```diff
@@ -0,0 +1,110 @@
+#region License
+
+// The PostgreSQL License
+//
+// Copyright (C) 2016 The Npgsql Development Team
+//
+// Permission to use, copy, modify, and distribute this software and its
+// documentation for any purpose, without fee, and without a written
+// agreement is hereby granted, provided that the above copyright notice
+// and this paragraph and the following two paragraphs appear in all copies.
+//
+// IN NO EVENT SHALL THE NPGSQL DEVELOPMENT TEAM BE LIABLE TO ANY PARTY
+// FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES,
+// INCLUDING LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS
+// DOCUMENTATION, EVEN IF THE NPGSQL DEVELOPMENT TEAM HAS BEEN ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//
+// THE NPGSQL DEVELOPMENT TEAM SPECIFICALLY DISCLAIMS ANY WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+// AND FITNESS FOR A PARTICULAR PURPOSE. THE SOFTWARE PROVIDED HEREUNDER IS
+// ON AN "AS IS" BASIS, AND THE NPGSQL DEVELOPMENT TEAM HAS NO OBLIGATIONS
+// TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+
+#endregion
+
+using System;
+using System.Linq.Expressions;
+using System.Security.Cryptography;
+using JetBrains.Annotations;
+using Remotion.Linq.Parsing.ExpressionVisitors.TreeEvaluation;
+
+namespace Npgsql.EntityFrameworkCore.PostgreSQL.Query.EvaluatableExpressionFilters.Internal
+{
+    /// <summary>
+    /// Represents an Npgsql-specific filter for hashing algorithms to identify expressions that are evaluatable.
+    /// </summary>
+    public class NpgsqlHashingAlgorithmEvaluatableExpressionFilter : IEvaluatableExpressionFilter
```

Comment:
Can you provide a bit more details? Why do we want this to always be evaluated at the server?

Reply:
> EF Core seemed to insist that the crypto methods evaluate locally.

Well, it does that because it can... Is there any reason you insist that they execute server-side?

In general, EF Core evaluates functions locally whenever it can, simply because there's no real reason not to. For example, if you call a function in a projection, there's no actual value in having it evaluated server-side rather than client-side. On the other hand, if you call a function in a WHERE clause, then there's definitely every reason to evaluate server-side (otherwise all the rows need to be pulled to the client).

The "evaluatable" filter is needed for cases where there is a special need, for a specific function, to *always* execute it server-side. There are several examples: random GUID generation will generate a different GUID every time, so the filter is used to make sure it's executed on each row's projection, rather than once, client-side, caching the results for all rows. Another good example is stuff like `DateTime.Now`, whose results you never want to reuse.

To summarize, I don't see anything special about these hash functions that would set them apart from other functions which aren't flagged is non-evaluatable client-side - am I missing something?